MAX32 RV32 (RISC-V) secondary core support

petejohanson-adi opened the pull request:
This PR adds support for running Zephyr on the secondary RV32 (RISC-V) core on the MAX32 parts that include it (MAX32655, MAX32690, MAX78002). The PR adds the necessary interrupt controller (custom) and tickless system timer (using one of the timer peripherals, MTIMER is not support on RV32) in order to fully support threading, etc.

The RV32 core is launched by the primary ARM core, so some small enablement for specifying the location for the RV32 address/entrypoint is added in order to do so.

Twister tests on the three mentioned targets have all been performed internally, including slow timer tests to ensure that implementation is sound.

A few items to point out, for reviewers, that impact other areas, and I welcome feedback on:
* The RV32 core does not support the `FENCE` instruction, so some minor changes to override that intrinsic where used in `lib/os/spsc_pbuf.c` are included.
* Flashing the RV32 core code is actually performed using a programmer attached to the ARM core (using openocd or jlink), which required some OpenOCD runner tweaks to allow specifying different pre-init to be used specifically when performing a flash operation. 

You can test locally with the following:

```
west build -p -b max32690evkit//m4 samples/sysbuild/hello_world -T sample.sysbuild.hello_world.max32690evkit_m4_rv32
```

And then flashing to an EVKit, and the corresponding code will be run on both cores. If you wish to run the Twister tests on the RV32 core, that same hello world sample is used as the "host program" on the ARM core to launch the RV32 code after a new test is flashed there, so before running twister, be sure to flash that sample to the evkit.

MaureenHelm committed:
dts: adi: max32: Move some devicetrees from arm to vendor

Moves some max32 devicetrees from arm to vendor to allow using them on
non-arm targets. This will make it possible to add support for the
secondary risc-v core that exists on these max32 variants.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm committed:
boards: adi: max32655evkit: Add support for risc-v core

Adds a new max32655evkit board target for the secondary risc-v core,

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm committed:
samples: sysbuild: Enable hello world for several MAX32 boards

Enables the sysbuild hello world sample application on MAX32 boards to
demonstrate using the soc's arm cortex-m4 and risc-v cores together. The
build system produces two separate application images, one for each core,
which can be flashed in two steps with the JLink runner.

1. Ensure you can access the serial console configured for the rv32 variant
   of your test board, e.g. setting JP4 and JP5 jumpers to connect LPUART
   to the USB UART connection CN1 on max32655evkit.

2. Build the sysbuild sample for your board's ARM board ID, which will
   automatically build the remote image for the board's RV32 variant, e.g.:

  $ west build -p -b max32655evkit//m4 samples/sysbuild/hello_world/ \
    -T sample.sysbuild.hello_world

3. Flash both domains at once with the JLink runner:

  $ west flash -r jlink

4. Hello world for the RV32 core will be printed to the serial console

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm committed:
soc: adi: max32: Add support for secondary risc-v core

Adds support for the secondary risc-v core on all applicable max32
family socs. The zero-riscy core supports the following instructions:

  - RV32I Base Integer Instruction Set
  - RV32E Base Integer Instruction Set (optional)
  - RV32C Standard Extension for Compressed Instructions
  - RV32M Integer Multiplication and Division Instruction Set Extension
    (optional)

Includes support for using pinctrl for setting up RV32 debug pin access,
and improved support for launcing the secodary core from the primary ARM
core.

See https://pulp-platform.org/docs/user_manual.pdf

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

petejohanson-adi committed:
boards: common: Use correct RV32 interface path

The olimex-arm-usb-ocd-h.cfg is under the fdti/ subdirectory.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

petejohanson-adi committed:
tests: kernel.common.stack_sentinel: Use RV32 illegal instruction

Use a custom illegal instruction for the MAX32 RV32 family, since it
doesn't trap on `unimp` or any other write to a read-only/non-existent CSR.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

petejohanson-adi committed:
drivers: hwinfo: Disable MAX32 hwinfo driver on RV32 core.

The flash info blocks needed to provide the device ID are not accessible to
the RV32 core on MAX32 devices, so disable this hwinfo driver there.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

petejohanson-adi committed:
tests: debug: Properly panic the MAX32 RV32 for coredump testing.

Explicitly `k_panic` on RV32 to test coredump on that target.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

petejohanson-adi committed:
drivers: serial: Disable async support on MAX32 RV32

MAX32 RV32 core DMA support is still a WIP, so disable async API for now
there.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

petejohanson-adi committed:
tests: subsys: secure_storage: Exclude max32 RV32 targets

Don't run the PSA ITS test on the max32 RV32 targets, since the automatic
erase there ends up erasing the flash region that contains the "host" ARM
application responsible for running the RV32 tests.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

MaureenHelm reviewed (dismissed):
Needs rebase

nordicjm commented on boards/adi/max32655evkit/doc/index.rst:
```diff
@@ -154,10 +154,73 @@ Debugging
+
+An experimental board configuration for the secondary RISC-V core can be found
+in
+:zephyr_file:`boards/adi/max32655evkit/max32655evkit_max32655_rv32_defconfig`.
```
@kartben for comment, would suggest deleting section. Board configuration isn't a hwmv2 term

nordicjm commented on boards/adi/max32655evkit/max32655evkit.dtsi:
```diff
@@ -0,0 +1,110 @@
+		rv32_sram: memory@20010000 {
+			compatible = "mmio-sram";
+			reg = <0x20010000 DT_SIZE_K(64)>;
+		};
```
this is not what reserved-memory is for

nordicjm commented on boards/adi/max32690evkit/max32690evkit.dtsi:
```diff
@@ -0,0 +1,135 @@
+		m4_sram: memory@20000000 {
+			compatible = "mmio-sram";
+			reg = <0x20000000 DT_SIZE_M(1)>;
+		};
```
as above, fix in whole PR

petejohanson-adi commented on soc/adi/max32/Kconfig:
```diff
@@ -72,13 +87,31 @@ config MAX32_STANDBY_DELAY
 config MAX32_SECONDARY_RV32_BOOT_ADDRESS
 	hex "Secondary RISC-V core boot address"
-	default $(dt_chosen_reg_addr_hex,$(DT_CHOSEN_Z_CODE_RV32_PARTITION))
+	default $(dt_chosen_reg_addr_hex,$(DT_CHOSEN_Z_RV32_FLASH)) if $(dt_chosen_enabled,$(DT_CHOSEN_Z_RV32_FLASH))
```
These values are used *by the ARM core* to know what address to set to launch the secondary RV32 core. The ARM core may have it's own code partition, but this is so the ARM core knows the code partition of the *other* core, so we need another value for that "launch the other core" objective.

petejohanson-adi commented on boards/adi/max32655evkit/doc/index.rst:
```diff
@@ -154,10 +154,73 @@ Debugging
+
+The primary Arm core uses Kconfig options,
+:kconfig:option:`CONFIG_MAX32_SECONDARY_RV32` and
+:kconfig:option:`CONFIG_MAX32_SECONDARY_RV32_BOOT_ADDRESS`, to enable and
```
Ok, changed this to pull from the `chosen` properties directly, and this made things a bit more flexible for targets with only one instance of the flash controller, where we can pull the base from `zephyr,flash` instead of needing a duplicate `zephyr,rv32-flash` chosen value.

Thoughts?

tejlmand commented on boards/adi/max32655evkit/max32655evkit_max32655_rv32_defconfig:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2023-2024 Analog Devices, Inc.
```
New file, is this copyright year correct ?

petejohanson-adi replied:
Fixed

tejlmand commented on tests/drivers/watchdog/wdt_basic_api/boards/max32_wdt.overlay:
Why is this file placed under `boards/` ?

The `boards/` folder is for files which should automatically be picked up by the build system and applied, based on the board target used.
It's not for files requiring users to specify `-DDTC_OVERLAY_FILE=...`

Should the file `max32_wdt.overlay` always be applied for the board target `max32655evkit/max32655/m4` ?

If yes, then you should name the file `max32655evkit_max32655_m4.overlay` instead.

petejohanson-adi replied:
It actually applies to multiple boards, which I've added to the list there for clarity.

tejlmand replied:
> It actually applies to multiple boards, which I've added to the list there for clarity.

I can see that, but that still doesn't mean it should go the the `boards` folder, as I wrote:
> The boards/ folder is for files which should automatically be picked up by the build system and applied,

Please move it out of the boards folder, and because the overlay looks quite generic (`zephyr,sram = &sram3`) then you could consider naming it according to it's purpose, `zephyr_sram3.overlay`, but I'll leave that decision up to you.

tejlmand commented on boards/common/openocd-adi-max32.boards.cmake:
```diff
@@ -22,7 +24,19 @@ board_runner_args(openocd --cmd-pre-init "source [find interface/${MAX32_INTERFA
 board_runner_args(openocd "--target-handle=_CHIPNAME.cpu")
 
+if(MAX32_FLASH_INTERFACE_CFG)
+  board_runner_args(openocd --flash-cmd-pre-init "source [find interface/${MAX32_FLASH_INTERFACE_CFG}]")
```
what would be the consequence of not having `--flash-cmd-pre-init` argument ?

petejohanson-adi replied:
See my comment here: https://github.com/zephyrproject-rtos/zephyr/pull/97309#discussion_r2662936885

tejlmand reviewed (changes requested):
Generally looks ok from build system perspective.

Only some minor areas for discussion.

ttmut commented on boards/adi/max78002evkit/max78002evkit_max78002_rv32.yaml:
```diff
@@ -0,0 +1,12 @@
+  - gpio
+  - serial
+ram: 112
+flash: 1024
```
Do you need to deduct this amount from `flash` in max78002evkit_max78002_m4.yaml?

ttmut commented on boards/adi/max78002evkit/max78002evkit.dtsi:
```diff
@@ -0,0 +1,171 @@
+		m4_storage_partition: partition@100000 {
+			label = "storage-m4";
+			reg = <0x100000 DT_SIZE_K(256)>;
+		};
```
Secure storage tests are failing for this board because they are failing to find a node with `storage_partition` label.

petejohanson-adi replied:
Fixed by added the label to the m4 board file. Thanks.

ttmut commented on dts/vendor/adi/max32/max32xxx-rv32.dtsi:
```diff
@@ -0,0 +1,89 @@
+	compatible = "adi,max32-rv32-sys-timer";
+	interrupts = <11 0>;
+	prescaler = <1>;
+	clock-source = <6>;
```
```suggestion
	clock-source = <ADI_MAX32_PRPH_CLK_SRC_ISO>;
```
Defined in `dt-bindings/clock/adi_max32_clock.h`

petejohanson-adi commented on scripts/west_commands/runners/openocd.py:
```diff
@@ -146,6 +148,10 @@ def do_add_parser(cls, parser):
+        parser.add_argument('--flash-cmd-pre-init', action='append',
+                            help='''Command to run before calling init when performing a flash;
+                            may be given multiple times;
+                            When set, existing --cmd-pre-init will be ignored when flashing''')
```
Ok, so here's the *need*, and I'm happy to move this to a separate issue/PR to discuss possible solutions:

Flashing via the RV32 debug connection is not possible(*) for MAX32 targets, so flashing code to the flash region we'll use for that core has to be orchestrated over the ARM core SWD connection.

On kits like the MAX78002-EVKIT, that has connections for jlink/MAX32625PICO to the ARM core, and another connection to use the included OLIMEX debugger to debug the RV32 core, you may have both connections active on your host, and want to easily `west flash` or `west debug` and have this "Just do the right thing" for each command.

To do so, we can override the flash operation to use a different config than the debug/attach one, so the flash operation is performed over the correct interface and with the correct target details. Having to manually specify the command every time you want to flash (or every time you want to debug/attach) is what this change is designed to avoid.

Footnote:

(*) in *theory* it may be possible for a very small subset of conditions where the ARM core is already set up to start the RV32 core, and is either not using the flash controller peripheral, or you're targeting one of the MAX32 variants with two flash controllers, but such conditions are rarely met, and not a solution for the need here generally.

petejohanson-adi commented on boards/adi/max32655evkit/doc/index.rst:
```diff
@@ -169,14 +169,6 @@ has significant limitations:
-
-- **Peripheral drivers**: Without interrupts, the existing UART, GPIO, and
-  other peripheral drivers won't work.
-
```
So... the interrupt controller is added later, to avoid the initial RV32 bring up landing as one huge commit. The original work (which was done before I started on the interrupt controller) had this limitation, which is why this is there, and was true at the time of that first commit.

I'm happy to just remove the lines in the *initial* commit, but that feels like an omission of details relevant to that particular state for later bisecting/history.

petejohanson-adi commented on drivers/timer/max32_rv32_sys_timer.c:
```diff
@@ -0,0 +1,200 @@
+
+uint32_t sys_clock_cycle_get_32(void)
+{
+	return MXC_TMR_GetCount(regs) * DT_INST_PROP(0, prescaler);
```
Per `include/zephyr/drivers/timer/system_timer.h` docs on the `sys_clock_cycle_get_32` function:

> The rate must match CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC.

And we're defaulting that on RV32 with:

```
config SYS_CLOCK_HW_CYCLES_PER_SEC
	default $(dt_node_int_prop_int,/clocks/clk_iso,clock-frequency)
```

So we need to match the ISO clock, which is why I believe the multiplier of the prescalar is required.

nordicjm reviewed (changes requested):
this PR needs to be reworked, there's 25 commits for many of them, I've looking at things that have just been added and are being removed or reworked, does not comply with zephyr contribution guidelines rule of no fixups

petejohanson-adi committed:
boards: adi: Add RV32 support to MAX32690EVKit

Add RV32 core target to the MAX32690EVkit board definition.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

petejohanson-adi committed:
soc: adi: max32: Add RV32 interrupt controller

Add interrupt controller for the RV32 core of the MAX32 SoC.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

petejohanson-adi committed:
drivers: timer: Add MAX32 RV32 system timer

The RV32 core does not implement RISC-V mtime, so add a new system
timer driver uses a peripheral timer for this functionality.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

petejohanson-adi committed:
board: adi: Add RV32 target to MAX78002EVKit board.

Add an additional target, max78002evkit/max78002/rv32 for targetting the
secondary RISC-V core on the board.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

nordicjm commented on dts/arm/adi/max32/max32660.dtsi:
```diff
@@ -43,6 +39,11 @@
 	soc {
+		sram0: memory@20000000 {
+			compatible = "mmio-sram";
+			reg = <0x20000000 DT_SIZE_K(16)>;
```
as above

nordicjm commented on dts/vendor/adi/max32/max32xxx-rv32.dtsi:
```diff
@@ -4,6 +4,26 @@
+
+		interrupt-parent = <&intc>;
+		intc: intc@e5070000 {
+			#address-cells = <1>;
```
not formatted properly, in fact I'm not sure how CI is not failing here

nordicjm commented on boards/adi/max32655evkit/max32655evkit.dtsi:
```diff
@@ -0,0 +1,106 @@
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
```
missing `ranges;`

nordicjm reviewed (changes requested):
flash changes are mandatory, the RAM ones are advised but optional

nashif reviewed (dismissed):
This PR does way too many things and ideally should be split into smaller manageable/reviewabled pieces. HW support PRs should not change core features/libs, if you nee to change something in lib or kernel/, submi this on its own in a smaller PR.

petejohanson-adi commented on lib/os/spsc_pbuf.c:
```diff
@@ -5,6 +5,7 @@
  */
 
 #include <zephyr/kernel.h>
+#include <soc.h>
```
I've removed the two patches related to this, so we can discuss how best to fix this in a dedicated PR/issue, and this PR can focus on the core changes for RV32 here. Thanks.

petejohanson-adi commented on dts/vendor/adi/max32/max78000-m4.dtsi:
```diff
@@ -0,0 +1,71 @@
+
+&flash0 {
+	interrupts = <23 0>;
+	reg = <0x10000000 DT_SIZE_K(512)>;
```
Fixed this up, and adjusted the `soc.c` startup code to account for the corrected address us to launch the RV32 core.

nordicjm reviewed (commented):
minor nit. then looks good

nordicjm reviewed (dismissed):
Board files OK
